pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -132,30 +146,23 @@ func newManagerImpl(socketPath string) (*ManagerImpl, error) {
 	return manager, nil
 }
 
-func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, added, updated, deleted []pluginapi.Device) {
-	kept := append(updated, added...)
+func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, devices []pluginapi.Device) {
 	m.mutex.Lock()
-	if _, ok := m.healthyDevices[resourceName]; !ok {
-		m.healthyDevices[resourceName] = sets.NewString()
-	}
-	if _, ok := m.unhealthyDevices[resourceName]; !ok {
-		m.unhealthyDevices[resourceName] = sets.NewString()
-	}
-	for _, dev := range kept {
+	m.healthyDevices[resourceName] = sets.NewString()
+	m.unhealthyDevices[resourceName] = sets.NewString()
+	for _, dev := range devices {
 		if dev.Health == pluginapi.Healthy {
 			m.healthyDevices[resourceName].Insert(dev.ID)
-			m.unhealthyDevices[resourceName].Delete(dev.ID)
 		} else {
 			m.unhealthyDevices[resourceName].Insert(dev.ID)
-			m.healthyDevices[resourceName].Delete(dev.ID)
 		}
 	}
-	for _, dev := range deleted {
-		m.healthyDevices[resourceName].Delete(dev.ID)
-		m.unhealthyDevices[resourceName].Delete(dev.ID)
-	}
 	m.mutex.Unlock()
 	m.writeCheckpoint()
+
+	if m.updateChan != nil {
+		m.updateChan <- new(interface{})
```

Comment:
This isn't used anywhere.

Reply:
thanks for review, this channel is added to help testing, test need to make sure callback been processed and then check stats, please refer to L109 at manager_test.go

Reply:
Ah I'm not sure that's a pattern we should use then. I'd rather you wrap that function in another callback in the test.
e.g:

```golang
originalCallback := m.callback
m.callback = func(...) ... {
     originalCallback(...)
     m.updateChan <- new(interface{})
}
```

Reply:
good idea, done, thanks